Making imports in core fully qualified

oscargus opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->

Related to  #22291

#### Brief description of what is fixed or changed

There were some non-fully qualified imports in core, making it hard to read the dependencies. Most of them are fixed.

I kept the `.`-imports, but removed the `..`-imports.

#### Other comments

I also moved the imports to a slightly more clever position regarding returns and exceptions before it was required.

This is not finished as some imports from `sympy` still exists (as I have not yet had time/energy to look up the locations).

#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* core
   * BREAKING CHANGE: `core.trace` with class `Tr` moved to `physics.quantum.trace`. 
<!-- END RELEASE NOTES -->

oscargus committed:
Making imports in core fully qualified

oscargus committed:
Moved core.trace to physics.quantum.trace

oscargus committed:
Fixed path of remaining imports

oscarbenjamin commented:
I've been through the diff and it all looks good except I would deprecate importing from `sympy.core.trace` rather than completely removing the module.

oscargus committed:
Added trace-stub back in core

oscargus commented:
I have added a stub with deprecation warning for `Tr` and updated the merge conflicts.

The previous comment is taken as an OK to merge, so I have enabled auto-merge...

smichr committed:
Update sympy/core/function.py

oscargus committed:
Fixed arg-test

oscargus committed:
Merge branch 'fullyqualifiedimportsincore' of github.com:oscargus/sympy into fullyqualifiedimportsincore

asmeurer commented on sympy/core/trace.py:
```diff
@@ -1,4 +1,10 @@
-from sympy import Expr, Add, Mul, Pow, sympify, Matrix, Tuple
+from .add import Add
+from .containers import Tuple
+from .expr import Expr
```
This deprecation decorator wasn't actually firing. The problem is that the @deprecated decorator was designed to add an `__init__` when added to a class. But because this class never returns itself, it never calls `__init__`. 

I'll look at making this work better, but a much simpler way to do this would be to just deprecate the entire module at import, and just add `from sympy.physics.quantum.trace import Tr`. The only issue with that is that module imports can only be run once, but that isn't a huge problem. 

I'm fixing this in https://github.com/sympy/sympy/pull/22900
